### PR TITLE
Don't fail on missing pip/certifi's cacert.pem

### PR DESCRIPTION
**Pull request by hroncok**
See https://github.com/pypa/pip/blob/master/src/pip/_vendor/README.rst

It is acceptable to remove the pip/_vendor/requests/cacert.pem file from pip

Virtualenv should not error on that

### CODE REVIEW

**Comment by hroncok** [comment]
Sure, tests are great.

What needs to be done in tests:

 * create a new virtual environment
 * install virtualenv package into it
 * patch the pip whl from the virtualenv_support package not to have and use certifi's cacert.pem but instead use `/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem'
 * create a new virtual environment using that virtualenv

this currentl blows with:

```
(__venv__) [virtualenvtest]$ python -m virtualenv test_virtualenv
Using base prefix '/usr'
New python executable in /home/churchyard/tmp/virtualenvtest/test_virtualenv/bin/python
Installing setuptools, pip, wheel...

  Complete output from command /home/churchyard/tmp...irtualenv/bin/python - setuptools pip wheel:
  Traceback (most recent call last):
  File "<stdin>", line 9, in <module>
  File "/usr/lib64/python3.7/pkgutil.py", line 637, in get_data
    return loader.get_data(resource_name)
OSError: [Errno 0] Error: 'pip/_vendor/certifi/cacert.pem'
----------------------------------------
...Installing setuptools, pip, wheel...done.
Traceback (most recent call last):
  File "/usr/lib64/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib64/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/churchyard/tmp/virtualenvtest/__venv__/lib64/python3.7/site-packages/virtualenv.py", line 2462, in <module>
    main()
  File "/home/churchyard/tmp/virtualenvtest/__venv__/lib64/python3.7/site-packages/virtualenv.py", line 762, in main
    symlink=options.symlink,
  File "/home/churchyard/tmp/virtualenvtest/__venv__/lib64/python3.7/site-packages/virtualenv.py", line 1015, in create_environment
    install_wheel(to_install, py_executable, search_dirs, download=download)
  File "/home/churchyard/tmp/virtualenvtest/__venv__/lib64/python3.7/site-packages/virtualenv.py", line 968, in install_wheel
    call_subprocess(cmd, show_stdout=False, extra_env=env, stdin=SCRIPT)
  File "/home/churchyard/tmp/virtualenvtest/__venv__/lib64/python3.7/site-packages/virtualenv.py", line 854, in call_subprocess
    raise OSError("Command {} failed with error code {}".format(cmd_desc, proc.returncode))
OSError: Command /home/churchyard/tmp...irtualenv/bin/python - setuptools pip wheel failed with error code 1
```

With my patch, it should not.

Now if I provide Python code that does and verifies the above, would you be able to guide me on where to stick that into the current test suite?

**Comment by gaborbernat** [comment]
Put it at the end of https://github.com/pypa/virtualenv/blob/master/tests/test_virtualenv.py#L334

**Comment by hroncok** [comment]
Added a test, verified that it fails without that fix.

File: tests/test_virtualenv.py [hunk 1]
```diff
@@ -332,3 +334,38 @@ def test_relative_symlink(tmpdir):
...
 
     assert os.path.exists(lib64)
+
+
+def test_missing_certifi_pem(tmpdir):
+    """Make sure that we can still create virtual environment if pip is
+    patched to not use certifi's cacert.pem and the file is removed.
+    This can happen if pip is packaged by Linux distributions."""
+    tests_dir = os.path.dirname(__file__)
+    support_original = os.path.join(tests_dir, "../src/virtualenv_support/")
```
Comment (gaborbernat) [inline tests/test_virtualenv.py]: this most certainly will fail on Windows; and indeed it has https://dev.azure.com/pypa/virtualenv/_build/results?buildId=3008 (not sure why the build does not show up here as Github check, although the job did start, you can see new jobs at https://dev.azure.com/pypa/virtualenv/_build?definitionId=11&_a=summary)

Comment (gaborbernat) [reply tests/test_virtualenv.py]: Please also add a changelog entry about the change while you are at it.

**Comment by hroncok** [comment]
OK, I have no idea what's the problem on Windows now or how to see any details :cry: 

Help me please?

File: tests/test_virtualenv.py [hunk 2]
```diff
@@ -332,3 +334,40 @@ def test_relative_symlink(tmpdir):
     shutil.rmtree(workdir)
 
     assert os.path.exists(lib64)
+
+
+def test_missing_certifi_pem(tmpdir):
```
Comment (gaborbernat) [inline tests/test_virtualenv.py]: use the new ``tmp_path`` please instead of ``tmpdir``. 👍

**Comment by hroncok** [comment]
oh I forgot that it goes through subprocess and is not working with forward slashes.

File: tests/test_virtualenv.py (hunk 2 above)
Comment (hroncok) [reply tests/test_virtualenv.py]: the test above uses tmpdir. what is tmp_path?

Comment (gaborbernat) [reply tests/test_virtualenv.py]: we'll migrate the old tests sometime in the future, but no point in not already use it for new tests - https://docs.pytest.org/en/latest/tmpdir.html?highlight=tmp_path#the-tmp-path-fixture

Comment (hroncok) [reply tests/test_virtualenv.py]: oh, pathlib! may i use pathlib also instead of the os.path functions?

File: tests/test_virtualenv.py [hunk 3]
```diff
@@ -332,3 +334,39 @@ def test_relative_symlink(tmpdir):
...
+    tests_dir = os.path.dirname(__file__)
+    proj_dir = os.path.dirname(tests_dir)
+    support_original = os.path.join(proj_dir, "src", "virtualenv_support")
+    pip_wheel = sorted(glob.glob(os.path.join(support_original, "pip*whl")))[0]
+    whl_name = os.path.basename(pip_wheel)
+
+    wheeldir = tmpdir / "wheels"
+    wheeldir.mkdir()
+    tmpcert = tmpdir / "tmpcert.pem"
+    cacert = "pip/_vendor/certifi/cacert.pem"
```
Comment (gaborbernat) [inline tests/test_virtualenv.py]: does not the same path separator counts on this and below (hence why still failing on Windows)?

**Comment by hroncok** [comment]
Windows still failing even with `os.path.join(proj_dir, "src", "virtualenv_support")`, no idea why. Also only on Python 3.

https://dev.azure.com/pypa/virtualenv/_build/results?buildId=3012&view=logs&jobId=debe4b1e-a6de-59f5-07ab-92771d5f9a9b&taskId=3601f2ed-d83f-5185-4e8b-01cc25e9c135&lineStart=42&lineEnd=42&colStart=38&colEnd=108

File: tests/test_virtualenv.py (hunk 3 above)
Comment (hroncok) [reply tests/test_virtualenv.py]: this is inside the zip but let me assert we found the file.

**Comment by gaborbernat** [comment]
Once we get the tests passing I would like to ask you to break up into a failing test only commit, and then another on top that does the source change and fixes it.

File: tests/test_virtualenv.py (hunk 3 above)
Comment (hroncok) [reply tests/test_virtualenv.py]: asserts work, different problem

**Comment by hroncok** [comment]
Azure now says Git fetch failed.

File: tests/test_virtualenv.py (hunk 2 above)
Comment (hroncok) [reply tests/test_virtualenv.py]: Done.

Comment (gaborbernat) [reply tests/test_virtualenv.py]: yeah 👍

**Comment by hroncok** [comment]
> Once we get the tests passing I would like to ask you to break up into a failing test only commit, and then another on top that does the source change and fixes it.

Shall I mark the test xfail in the first commit or not?

----

I have figured out that if the script I've changed fails, there is no error output anywhere in the test log, so I've temporarily enabled that output for now to see what's wrong on the Windows Python 3 code (maybe it's juts a different exception type). However the CI won't start.

**Comment by gaborbernat** [comment]
No need for xfail just let it fail. if you ammend your commit it my trigger a redirect.

**Comment by gaborbernat** [comment]
https://dev.azure.com/pypa/virtualenv/_build/results?buildId=3021

File: tests/test_virtualenv.py [hunk 4]
```diff
@@ -332,3 +339,44 @@ def test_relative_symlink(tmpdir):
...
+    pip_wheel = sorted(support_original.glob("pip*whl"))[0]
+    whl_name = pip_wheel.name
+
+    wheeldir = tmp_path / "wheels"
+    wheeldir.mkdir()
+    tmpcert = tmp_path / "tmpcert.pem"
+    cacert = "pip/_vendor/certifi/cacert.pem"
+    certifi = "pip/_vendor/certifi/core.py"
+    oldpath = b"os.path.join(f, 'cacert.pem')"
+    newpath = "r'{}'".format(tmpcert).encode("ascii")
```
Comment (hroncok) [inline tests/test_virtualenv.py]: So the Windows path has backslashes and if you put `\u` into str, it blows up :grinning:

**Comment by gaborbernat** [comment]
thanks

**Comment by hroncok** [comment]
Thanks to you.

File: tests/test_virtualenv.py (hunk 4 above)
Comment (gaborbernat) [reply tests/test_virtualenv.py]: easy to achieve by running in c:\users folder

**Pull request merged**